### PR TITLE
[Feat] UI Components - Customize form field on authentication components

### LINKED ISSUE #5298: UI Components - AmplifyAuthenticator - How to use formFields for Sign Up? [React]
** Which Category is your question related to? **
Auth

** What AWS Services are you utilizing? **
Amplify, Cognito, API GW, Lambda

** Provide additional details e.g. code snippets **
I want to change the fields shown in the Sign Up form. Currently Username, Password, Email and Phone Number is displayed. I only need to get Email and Password. How can I change SignUp form fields?

I was able to do it by providing signUpConfig parameter while using withAuthenticator; however with AmplifyAuthenticator I could not make it. I tried to use formFields property as below; however it didn't work.
```
import React from "react";
import Amplify from "aws-amplify";
import { AmplifyAuthenticator, AmplifySignIn, AmplifySignUp, AmplifySignOut } from "@aws-amplify/ui-react";
import awsconfig from "./aws-exports";

Amplify.configure(awsconfig);

const App = () => (
  <AmplifyAuthenticator usernameAlias="email">
    <AmplifySignIn
      headerText="Sign In Header"
      slot="sign-in" />
    <AmplifySignUp headerText="Sign Up Header"
      formFields={[{ type: 'email', label: 'Email', placeholder: 'Email', hint: 'Enter Your Email', required: true }, { type: 'password', label: 'Password', placeholder: 'Enter Your Password', hint: 'Password', required: true }]}
      slot="sign-up" />

    <div>
      Internal App
      <AmplifySignOut />
    </div>
  </AmplifyAuthenticator>
);

export default App;
```

### PR DESCRIPTION
**Pull request by ashika01**
Fixes #5298 

**_Description of changes:_**
- Customize the form fields using slots
- Add support for `userAttributes` for signUp 

**_Update:_** Got the customization working with form fields. Need update for supporting all standard and custom user attributes.

Screen Shot 2020-04-15 at 3 09 11 AM

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-ui-components/src/components/amplify-auth-fields/amplify-auth-fields.tsx [hunk 1]
```diff
@@ -30,9 +30,17 @@ export class AmplifyAuthFields {
...
     if (formFields === undefined) return '';
 
-    formFields.forEach((formField: FormFieldType | string) =>
-      content.push(componentFieldMapping[typeof formField === 'string' ? formField : formField.type](formField)),
-    );
+    formFields.forEach((formField: FormFieldType | string) => {
+      if (typeof formField === 'string') {
+        content.push(componentFieldMapping[formField](formField));
+      } else {
+        if (Object.keys(componentFieldMapping).includes(formField.type)) {
```
Comment (jordanranz) [inline packages/amplify-ui-components/src/components/amplify-auth-fields/amplify-auth-fields.tsx:37]: Let's move this line to an `else if`

File: packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx [hunk 2]
```diff
@@ -49,36 +49,62 @@ export class AmplifyConfirmSignUp {
...
-            <amplify-button variant="anchor" onClick={() => this.resendConfirmCode()}>
-              {I18n.get(Translations.CONFIRM_SIGN_UP_RESEND_CODE)}
-            </amplify-button>
-          </div>
-        ),
-        handleInputChange: event => this.handleCodeChange(event),
-      },
-    ];
+    if (this.formFields.length === 0) {
+      this.formFields = [
```
Comment (jordanranz) [inline packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx:57]: Is it possible to keep the `formFields` prop read-only here and just set the local `newFormFields` to this list. Especially if we are passing  `newFormFields` to the auth-fields component

Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx:57]: I think this applies to all components in this PR utilizing `formFields`^

File: packages/amplify-ui-components/src/components/amplify-forgot-password/amplify-forgot-password.tsx [hunk 3]
```diff
@@ -41,68 +41,95 @@ export class AmplifyForgotPassword {
...
+      default:
+        return;
+    }
   }
 
-  handleEmailChange(event) {
-    this.email = event.target.value;
+  handleFormFieldInputWithCallback(event, field) {
+    let fnToCall = field['handleInputChange'];
+    let callback =
```
Comment (jordanranz) [inline packages/amplify-ui-components/src/components/amplify-forgot-password/amplify-forgot-password.tsx:128]: These can be `const` if we are not resetting them

File: packages/amplify-ui-components/src/components/amplify-sign-up/amplify-sign-up.tsx [hunk 4]
```diff
@@ -116,143 +130,147 @@ export class AmplifySignUp {
...
+        this.signUpAttributes.username = this.signUpAttributes.attributes.email;
         break;
       case 'phone_number':
-        this.userInput = composePhoneNumberInput(this.phoneNumber);
+        this.signUpAttributes.username = composePhoneNumberInput(this.phoneNumber);
         break;
       case 'username':
       default:
-        this.userInput = this.username;
+        this.signUpAttributes.username = this.signUpAttributes.username;
```
Comment (jordanranz) [inline packages/amplify-ui-components/src/components/amplify-sign-up/amplify-sign-up.tsx:140]: Is this assignment needed?

File: packages/amplify-ui-components/src/components/amplify-forgot-password/amplify-forgot-password.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashika01) [reply packages/amplify-ui-components/src/components/amplify-forgot-password/amplify-forgot-password.tsx:128]: yup I can resolve this.

File: packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx (hunk 2 above)
Comment (ashika01) [reply packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx:57]: Talked offline. Cleaning up to newFormFields for this instance. But we need deeper thinking on this prop being readonly.

**Pull request merged**